### ISSUE TITLE
cProfiling geoplot vs Cartopy vs GeoPandas: puzzling results

### ISSUE THREAD

**Issue by gregorhd** [ResidentMario/geoplot#250]
Hi there,

apologies in advance for the long post. I’m currently [comparing](https://github.com/gregorhd/mapcompare), among others, different geospatial mpl-interfaces including _cartopy_, _GeoPandas_, and _geoplot_ for a thesis. One part is comparing the total CPU runtimes of each library to generate the same map product, using the same data. The performance benchmarking is done with _cProfile_.

**The basic setup is this**: After the data has been prepared and transformed to where each of the libraries can handle it directly, I’m wrapping all figure and axes definitions, including the central `gplt.polyplot()` calls (or `.plot()` for GeoPandas/ `add_geometries` for _Cartopy_) in a function called `renderFigure()` which is then wrapped with a decorator that basically does this:

```
p = cProfile.Profile() 
p.enable() # start cProfiling

value = func(*args, **kwargs) # execute renderFigure()

p.disable() # stop cProfiling
p.dump_stats() # write cProfile to file
```

After a tip over in the _Cartopy_ Gitter channel when I noticed _Cartopy_ showing unrealistically low runtimes, it turned out that, due to lazy execution of the underlying rendering functions in some libraries, an explicit call to `fig.canvas.draw()` caused the results for _Cartopy_ and _GeoPandas_ looking a lot more comparable. What changed, I presume, was when _cProfile_ considered `renderFigure()` to be 'finished'. You can see the difference in _Cartopy_ runtimes on the large 144,000 polygon dataset below before and after adding the forced draw (in seconds).

![image](https://user-images.githubusercontent.com/75640459/130054524-2b0914e7-d1da-49f4-8dd7-af20ad63174f.png)

Now with _geoplot_, the situation is also a bit tricky. If I only force a `draw()` for _Cartopy_ and _GeoPandas_ (since no `fig` object is created when using geoplot's standard syntax), _geoplot_ compares extremely favourably to the others making it 30% faster than _Cartopy_'s and 40% faster than GeoPandas' _mpl_ interface.

![image](https://user-images.githubusercontent.com/75640459/130062204-4fbde3d3-da34-4b2e-99b0-361a1c1085c7.png)

But I'm starting to question whether this makes sense, given that _geoplot_ relies on the same plotting backend, namley _Matplotlib_, and thus shouldn't be that much more performant.

These results also jar with what I'm actually observing when using _geoplot_: _cProfile_ would complete fairly quickly (a print statement tells me when the dump to `.prof` has finished) but the actual figure would not appear in the interpreter until much, much later compared with _Cartopy_ or _GeoPandas_. So I tried to understand what's happening behind the scenes to see whether 'lazy execution' was also going on here and if I needed to also force the canvas draw for _geoplot_ so that I'm comparing like with like. Below my initial, amateur investigation:

A standard call like `ax = gplt.polyplot()` I believe triggers [geoplot.py line 675](https://github.com/ResidentMario/geoplot/blob/da6e7ecb906b0ac988d0063aae1434e02071ec84/geoplot/geoplot.py#L675):
```
if self.ax is None:
   ax = plt.gca()
```
leading to [matplotlib.pyplot line 2491](https://github.com/matplotlib/matplotlib/blob/64619e53e9d0ed417daba287ac0d3a06943a54d5/lib/matplotlib/pyplot.py#L2491):
```
def gca(**kwargs):
    return gcf().gca(**kwargs)
```
then to [matplotlib.pyplot line 831](https://github.com/matplotlib/matplotlib/blob/64619e53e9d0ed417daba287ac0d3a06943a54d5/lib/matplotlib/pyplot.py#L831)

`return figure()`

and, not quite ultimately but close, to [matplotlib.figure line 1508](https://github.com/matplotlib/matplotlib/blob/64619e53e9d0ed417daba287ac0d3a06943a54d5/lib/matplotlib/figure.py#L1508):

```
if self._axstack.empty():
   return self.add_subplot(1, 1, 1, **kwargs)
```
which then, with some additional steps, adds an Axes to the figure as part of a subplot arrangement. This is sort of where I'm starting to lose the thread, not being a Python dev at all.

I'm still no closer to understanding why figures take so long to draw with _geoplot_ and why the runtimes diverge so much from the other _mpl_-based libraries. Following the cookie trail for _cartopy_'s `add_geometries()` function, for instance, it turns out that it hooks into the `FeatureArtist` class of _matplotlib_, so there are clearly different approaches being used though without having the devs explain the reasoning behind the choices it's tough to reverse engineer it from the outside.

In an attempt to level the playing field, I tried [adding ](https://github.com/gregorhd/mapcompare/blob/afc7eea9d81b8f1dd9709c18ba14845e771fd340/scripts/gplt.py#L106) a `fig` specifically via `fig = plt.gcf()` to my `renderFigure()` function for _geoplot_ so I could call then call `fig.canvas.draw()` on it. The results I get then _roughly_ match up with what I'm seeing in terms of how long it takes for figures to show up (it really takes that long), as shown below:

![image](https://user-images.githubusercontent.com/75640459/130060495-7229471e-fe81-41e0-9a48-665b55107cf0.png)

But I'm not sure a) whether this result is any more fair than the first, or b) whether I'm inadvertantly skewing the results by trying to establish a 'level playing field'. I'd hate to give readers (at most two I'm guessing...) the wrong impression, especially with _geoplot_ otherwise doing extremely well in terms of code complexity and userfriendliness.

So, long story short. If you guys have any thoughts on what could be going on, that'd be much appreciated! I can understand though, if digging into the weeds on these things (especially trying to understand the massive Gordian knot which is _Matplotlib_) simply takes up too much time.

Cheers,
Gregor

**Comment by ResidentMario** [comment]
I'm afraid I don't have any special insight to offer.

`matplotlib` and `cartopy` are both very complex library. `matplotlib`, dating to the era when Python was new, is not particularly well-designed from a modern perspective; `cartopy` IIRC makes use of baroque C stuff like `gdal`. `geoplot` does not do very much work on its own (with a couple of exceptions), so understanding its perf is mostly a matter of understanding the perf of these two libraries it wraps, which is no small task.

`geoplot` is a side project of mine, something that I try to maintain (and fix bugs when people report them) but which is extremely irrelevant to my day job and hence not something I can afford to spend much time on. Performance is, for the most part, not something I have control over, nor something I really prioritized (otherwise I wouldn't be using `matplotlib`!), so I mostly haven't been paying attention.

**Comment by gregorhd** [comment]
Hi Aleksey, thanks for the context! That's completely understandable. I'll try to do some more digging and will report back if I come across anything interesting. But rest assured that the performance section will be accompanied by a good few caveats anyway :)
